### PR TITLE
Clarify buffer creation with nullptr

### PR DESCRIPTION
**Pull request by gmlueck**
Clarify that buffer and image constructors that take `shared_ptr`
support an empty `shared_ptr`, which causes the buffer or image to have
uninitialized contents.  There was previously a statement to this
effect in section 4.7.4.3 "Shared SYCL ownership of the host memory",
but the sections describing the constructors themselves didn't say
anything about this.

Null raw pointers are still not allowed, consistent with the previous
wording.  However, I clarified that null raw pointers are allowed when
the buffer size is also zero.

I also clarified the constructors taking `shared_ptr` to note that the
application can safely release ownership of the pointer after the
constructor returns.  This is consistent with the existing wording and
example in section 4.7.4.3 "Shared SYCL ownership of the host memory".

In order to avoid confusion, I changed wording like:

> The ownership of this memory is given to the constructed SYCL buffer

To:

> The buffer assumes exclusive access to this memory

This seems better because the word "ownership" has a different meaning
for `shared_ptr`.

### CODE REVIEW

**Comment by gmlueck** [comment]
There is no issue for this PR, so we haven't discussed this yet.  Feel free to object if you don't agree.

I do think we want the behavior of all the buffer constructors which take a `hostData` pointer to be consistent w.r.t. null.  The other option is to allow null also for the overloads that take a raw pointer.  I tend to think that the statement I removed about allowing empty `shared_ptr` was just an oversight because it wasn't in the main description of the buffer constructors.  If we really intended to allow that, I think we would have made it clearer in the description of the constructor.

**Comment by tomdeakin** [comment]
Can you remind me where it's pretty clear that a `nullptr` raw pointer is not allowed? I can't find that bit of the spec right now... The only place I can find is in the description of the `const T *` buffer constructor which says "another valid non-null final data address specified via the member function `set_final_data()`" which could by implication mean the `hostData` pointer needed to be non-null. We don't say this for the non-const version of the constructor though.

We need to know what happens in that case, as the paragraph deleted still says a buffer allocates memory even if the `shared_ptr` was empty. I think I would expect a null raw pointer would still allocate non-zero sized the buffer data, but just not initialise it: i.e. the constructor would have the same behaviour as the one that didn't take a host pointer. This is different behaviour to what is changed in this PR, so I think we probably need some discussion.

**Comment by gmlueck** [comment]
> Can you remind me where it's pretty clear that a nullptr raw pointer is not allowed? I can't find that bit of the spec right now...

I think it is clear by omission.  No where does the spec say that `nullptr` is allowed or that the buffer constructor has any special behavior when the pointer is null.  The spec currently says this for the constructors that take a raw `hostData` pointer:

> Construct a SYCL `buffer` instance with the `hostData` parameter provided. The buffer is initialized with the memory specified by `hostData`.  The ownership of this memory is given to the constructed SYCL buffer for the duration of its lifetime.

The statement "The buffer is initialized with the memory specified by `hostData`" certainly implies that `hostData` points to some valid data.  The statement "The ownership of this memory is given to the constructed SYCL buffer for the duration of its lifetime" implies that `hostData` points to some memory.

If we intended to allow null here, I think the spec would have said something like "If `hostData` is `nullptr`, this constructor is equivalent to `buffer(const range<Dimensions>& bufferRange, const property_list& propList = {})`".  But, we didn't say that.

File: adoc/chapters/programming_interface.adoc [hunk 1]
```diff
@@ -4241,6 +4241,11 @@ merely allocate uninitialized memory, and the application is responsible for
 constructing objects by calling placement-new and for destroying them later
 by manually calling the object's destructor.
 
+For the overloads that do copy objects from host memory, the [code]#hostData#
+pointer must point to _N_ bytes of memory where _N_ is
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc]: Done in 9c9399307636808202b05f188c528e70a6d29629

**Comment by keryell** [comment]
After long discussion during F2F about the blocking semantics and unification of shared_ptr and raw pointer semantics, it looks like there is an agreement about converging on the same effect during destruction which is that if there is no copy back, it is non blocking.
We agree that this should be simplified in SYCL Next but, for now, do not change the SYCL 2020 semantics.
So, do not merge this for now and come with a new version of the PR.

**Comment by gmlueck** [comment]
> So, do not merge this for now and come with a new version of the PR.

I have updated this PR, and I believe it is now ready for re-review and merge.

**Comment by keryell** [comment]
Greg thinks it is ready to go but we need some reviews.
Discussed during last F2F.

**Review by nliber — APPROVED** [review]

**Review by TApplencourt — APPROVED** [review]

**Review by keryell — APPROVED** [review]
Thanks for the clarifications.

**Review by AerialMantis — COMMENTED** [review]
Looks good, just one comment on the wording about different constructors.

File: adoc/chapters/programming_interface.adoc [hunk 2]
```diff
@@ -4188,6 +4188,11 @@ merely allocate uninitialized memory, and the application is responsible for
 constructing objects by calling placement-new and for destroying them later
 by manually calling the object's destructor.
 
+For the overloads that do copy objects from host memory, the [code]#hostData#
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc:4191]: Yes, I meant the constructor overloads.  I think we could reword this paragraph to be clearer in general, though.  For example, we need to make it clear that it only applies to the constructors that take a pointer argument, not to other constructors that copy objects from host memory (e.g. the overload that takes a container).  How about this:

> For the overloads of the `buffer` constructor that take a raw pointer to the `T` type, the pointer must must point to at least _N_ bytes of memory where _N_ is `sizeof(T) * bufferRange.size()`.  If _N_ is zero, the pointer is permitted to be null.  For the overloads of the `buffer` constructor that take an `std::shared_ptr` to `T` or `T[]`, the pointer must either be empty or it must point to at least _N_ bytes of memory where _N_ is `sizeof(T) * bufferRange.size()`.

**Pull request merged**